### ISSUE TITLE
[MNG-4452] Metadata for snapshots should include classifier

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#5719]
**[Paul Benedict](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=paul4christ79)** opened **[MNG-4452](https://issues.apache.org/jira/browse/MNG-4452?redirect=false)** and commented

Please see the whole conversation here:
http://old.nabble.com/Multi-Platform-snapshots-td25295999.html

Essentially, an artifact's classifier isn't included in the repository's metadata of a snapshot. This makes it difficult (impossible?) to deploy multiply snapshots with different classifiers. The primary use case is to continuously build an artifact for different environments.

---

**Affects:** 2.2.1, 3.0-alpha-4

**Attachments:**
- [nar-artifact-list.txt](https://issues.apache.org/jira/secure/attachment/12713050/nar-artifact-list.txt) (_102.65 kB_)
- [nar-maven-metadata.xml](https://issues.apache.org/jira/secure/attachment/12712409/nar-maven-metadata.xml) (_1.73 kB_)
- [output](https://issues.apache.org/jira/secure/attachment/12713249/output) (_232.50 kB_)

**Issue Links:**
- [MNG-3408](https://issues.apache.org/jira/browse/MNG-3408) Artifacts with the same pair groupId:artifactId but different type are not resolved independently
 (_**"is duplicated by"**_)
- [MDEPLOY-48](https://issues.apache.org/jira/browse/MDEPLOY-48) deploy:deploy-file does not support deploying sources jars too

- [MNG-4965](https://issues.apache.org/jira/browse/MNG-4965) resolver fails to find artifacts with classifiers from previous deployments
 (_**"is depended upon by"**_)

3 votes, 6 watchers

**Comment by jira-importer** [comment]
**[Dan Tran](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dantran)** commented

don't think 2.2.1 is affected. Just 3.0

**Comment by jira-importer** [comment]
**[Paul Benedict](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ID134349)** commented

Dan, can you please check to verify?

**Comment by jira-importer** [comment]
**[Dan Tran](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dantran)** commented

confirm, it only affects maven 3

**Comment by jira-importer** [comment]
**[Benjamin Bentmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bentmann)** commented

Not sure what exactly you consider as being affected, the metadata used by Maven 2 and 3 is the same. Maven 3.0 does not support non-unique snapshots but the general problem with the metadata exists in all existing Maven versions.

**Comment by jira-importer** [comment]
**[Dan Tran](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dantran)** commented

sorry about the confusion. This issue should mainly raise concern where maven 3 removes non-unique snapshot feature, and does not provide an alternative.

deploy non-unique snapshot still works with 2.2.1

**Comment by jira-importer** [comment]
**[Benjamin Bentmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bentmann)** commented

Fixed in [r995057](http://svn.apache.org/viewvc?view=revision&revision=995057).

**Comment by jira-importer** [comment]
**[Benjamin Bentmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bentmann)** commented

To support interop with existing repository managers that don't know about the new format but try to process it on the server side, the system property `maven.metadata.legacy=true|false` has been added in [r995081](http://svn.apache.org/viewvc?view=revision&revision=995081) to revert to the 1.0.0 format during deployment if needed.

**Comment by jira-importer** [comment]
**[Andrei Pozolotin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=andrei.pozolotin)** commented

Benjamin:
just to confirm, is it fixed in 3.0.1? the issue still affecting me
thanks;
Andrei

**Comment by jira-importer** [comment]
**[Andrei Pozolotin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=andrei.pozolotin)** commented

I guess, I must clarify:

1) yes I see in 3.0.1 that shapshots metadata does includes classifier;

2) the problem is that resolver still does not find attached artifacts form previous deployments;

for example, this is repo metadata after 2 deployments:

note, that artifacts of interest are (build 83 and 84):

com.barchart.udt:barchart-udt4:1.0.2-20110108.023638-83:i386-Linux-g++-jni:nar
com.barchart.udt:barchart-udt4:1.0.2-20110108.023723-84:amd64-Linux-g++-jni:nar

##############################################################

\<metadata modelVersion="1.1.0">
\<groupId>com.barchart.udt\</groupId>
\<artifactId>barchart-udt4\</artifactId>
\<version>1.0.2-SNAPSHOT\</version>
&#8722;
\<versioning>
&#8722;
\<snapshot>
\<timestamp>20110108.023723\</timestamp>
\<buildNumber>84\</buildNumber>
\</snapshot>
\<lastUpdated>20110108023723\</lastUpdated>
&#8722;
\<snapshotVersions>
&#8722;
\<snapshotVersion>
\<extension>jar\</extension>
\<value>1.0.2-20110108.023723-84\</value>
\<updated>20110108023723\</updated>
\</snapshotVersion>
&#8722;
\<snapshotVersion>
\<extension>pom\</extension>
\<value>1.0.2-20110108.023723-84\</value>
\<updated>20110108023723\</updated>
\</snapshotVersion>
&#8722;
\<snapshotVersion>
\<classifier>noarch\</classifier>
\<extension>nar\</extension>
\<value>1.0.2-20110108.023723-84\</value>
\<updated>20110108023723\</updated>
\</snapshotVersion>
&#8722;
\<snapshotVersion>
\<classifier>i386-Linux-g++-jni\</classifier>
\<extension>nar\</extension>
\<value>1.0.2-20110108.023638-83\</value>
\<updated>20110108023638\</updated>
\</snapshotVersion>
&#8722;
\<snapshotVersion>
\<classifier>amd64-Linux-g++-jni\</classifier>
\<extension>nar\</extension>
\<value>1.0.2-20110108.023723-84\</value>
\<updated>20110108023723\</updated>
\</snapshotVersion>
\</snapshotVersions>
\</versioning>
\</metadata>

##############################################################

and this is dependency declaration, I want 1 jar and 2 nar, platform dependent:

##############################################################

    	<dependency>
    		<groupId>com.barchart.udt</groupId>
    		<artifactId>barchart-udt4</artifactId>
    		<version>${barchart.version}</version>
    		<type>jar</type>
    	</dependency>
    
    	<dependency>
    		<groupId>com.barchart.udt</groupId>
    		<artifactId>barchart-udt4</artifactId>
    		<version>${barchart.version}</version>
    		<classifier>i386-Linux-g++-jni</classifier>
    		<type>nar</type>
    	</dependency>
    
    	<dependency>
    		<groupId>com.barchart.udt</groupId>
    		<artifactId>barchart-udt4</artifactId>
    		<version>${barchart.version}</version>
    		<classifier>amd64-Linux-g++-jni</classifier>
    		<type>nar</type>
    	</dependency>

##############################################################

and this is the error:

##############################################################

[INFO] Scanning for projects...
[INFO] snapshot com.barchart.udt:barchart-udt4-parent:1.0.0-SNAPSHOT: checking for updates from sonatype-nexus-snapshots
[INFO] ------------------------------------------------------------------------
[INFO] Building Unnamed - com.barchart.udt:barchart-udt4-bundle:jar:1.0.0-SNAPSHOT
[INFO]    task-segment: [validate]
[INFO] ------------------------------------------------------------------------
[INFO] snapshot com.barchart.udt:barchart-udt4:1.0.2-SNAPSHOT: checking for updates from sonatype-nexus-snapshots
Downloading: https://oss.sonatype.org/content/repositories/snapshots/com/barchart/udt/barchart-udt4/1.0.2-SNAPSHOT/barchart-udt4-1.0.2-20110108.023723-84.pom

Downloading: https://oss.sonatype.org/content/repositories/snapshots/com/barchart/udt/barchart-udt4/1.0.2-SNAPSHOT/barchart-udt4-1.0.2-20110108.023723-84.jar
78K downloaded  (barchart-udt4-1.0.2-20110108.023723-84.jar)
Downloading: https://oss.sonatype.org/content/repositories/snapshots/com/barchart/udt/barchart-udt4/1.0.2-SNAPSHOT/barchart-udt4-1.0.2-20110108.023723-84-i386-Linux-g++-jni.nar
Downloading: https://oss.sonatype.org/content/repositories/snapshots/com/barchart/udt/barchart-udt4/1.0.2-SNAPSHOT/barchart-udt4-1.0.2-20110108.023723-84-amd64-Linux-g++-jni.nar

[INFO] ------------------------------------------------------------------------
[ERROR] BUILD ERROR
[INFO] ------------------------------------------------------------------------
[INFO] Failed to resolve artifact.

##############################################################

in other words, the resolver tries to find

barchart-udt4-1.0.2-20110108.023723-84-i386-Linux-g++-jni.nar
and
barchart-udt4-1.0.2-20110108.023723-84-amd64-Linux-g++-jni.nar

both from latest build #84, instead of looking through the versioning
in order to find the file that actually exists in build #83:

barchart-udt4-1.0.2-20110108.023638-83:i386-Linux-g++-jni.nar

**Comment by jira-importer** [comment]
**[Andrei Pozolotin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=andrei.pozolotin)** commented

forgot to mention: of course, this relates to SNAPSHOT artifacts only

**Comment by jira-importer** [comment]
**[Leon Franzen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=leon_franzen)** commented

We are seeing exactly the same problem as Andrei in Maven 3.0.1 with platform dependent nar builds.  This is a critical blocker for us.

By reconfiguring our settings.xml to work around the inability of Nexus to provide the new classifier/type specific metadata information through a repository-group, we are now stuck on this issue.  Is there a way to set the build number at deploy time to work around this issue?

**Comment by jira-importer** [comment]
**[Leon Franzen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=leon_franzen)** commented

The state of the snapshot repository during the Maven build shown in the file "output" (attached)

**Comment by jira-importer** [comment]
**[Andrei Pozolotin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=andrei.pozolotin)** commented

Leon:
my woraround is use direct repo access and NOT to use repo-to-repo proxy:
https://issues.sonatype.org/browse/NEXUS-4055
Andrei

**Comment by jira-importer** [comment]
**[Leon Franzen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=leon_franzen)** commented

Then I think perhaps I am not configuring my settings.xml properly to work around the problem.  I am still referencing a repository group but have tried to include a direct repo reference to my snapshot repository.

\<settings>
\<mirrors>
\<mirror>
\<id>nexus\</id>
\<mirrorOf>*,!snapshots\</mirrorOf>
\<url>http://nexus:8081/nexus/content/groups/public\</url> \
\</mirror>
\</mirrors>
\<profiles>
\<profile>
\<id>nexus\</id>
\<repositories>
\<repository>
\<id>snapshots\</id>
\<url>http://nexus:8081/nexus/content/repositories/snapshots\</url>  \
\<releases>
\<enabled>false\</enabled>
\</releases>
\<snapshots>
\<enabled>true\</enabled>
\</snapshots>
\</repository>
...
Andrei, how were you specifically able to work around the issue?  Are you not using a group?

**Comment by jira-importer** [comment]
**[Andrei Pozolotin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=andrei.pozolotin)** commented

Leon:

in my case:

1) my "real repo" is sonaytpe oss;

2) my "proxy repo" is internal corporate repo which was going after sonatype oss;

Andrei

**Comment by jira-importer** [comment]
**[Andrei Pozolotin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=andrei.pozolotin)** commented

Leon:

according to this:
https://issues.sonatype.org/browse/NEXUS-4055

sonatype folks fixed this in trunk; you may want to build nexus;

Andrei

**Comment by jira-importer** [comment]
**[Andrei Pozolotin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=andrei.pozolotin)** commented

Leon:

I tried to send you email (listed on this site); but it got bounced;

I wanted to ask you if you were able to solve the following nar issue:
https://issues.sonatype.org/browse/NAR-181

thank you;

Andrei

**Comment by jira-importer** [comment]
**[Lukasz Szelag](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=lukasz74nj)** commented

I'm also experiencing the same issue (Maven 2.2.1 and Archiva 1.3.5). As I'm not ready to switch to Maven 3.x yet - is there a patch for 2.2.1 that provides a fix? I'm using this functionality (classifier) to name artifacts for different environments.

Thanks,
Lukasz
